Make "loaded_link" and "no_results_link" more configurable

dsteelma-umd opened the pull request:
The "loaded_link" provides a link to the native search interface,
with the current search pre-populated. The "no_results_link" provides
a link to the native search interface when no search results are found.

In the sample searchers, such as "quick_search-placeholder_searcher"
(https://github.com/NCSU-Libraries/quick_search-placeholder_searcher)
the "loaded_link" is defined both in the I18N locale file
(config/locales/en.yml file), and in a "loaded_link" method in the
implementation of the Searcher base class
(app/searchers/quick_search/placeholder_searcher.rb)

So in the GUI, can be "loaded_link" is retrieved from two different
places:

1) The "loaded_link" property from the I18N locale file is used in the
following view helpers:

* quick_search/app/views/quick_search/search/_search_error.html.erb
* quick_search/app/views/quick_search/search/_no_results.html.erb

2) The "loaded_link" method from the searcher class is used in the
"quick_search/app/views/quick_search/search/_module.html.erb" view
helper.

Having a "loaded_link" property in the en.yml file, and a "loaded_link"
method in the searcher is confusing, as it is unclear without examining
the code, which instance is used where. It is even more problematic,
because the return value from the "loaded_link" method may contain
additional processing that is not possible with the "loaded_link"
property.

The I18N locale files also have a "no_results_link" property. There is
no associated "no_results_link" method on the searcher, so this property
cannot always reflect the same URL as the "loaded_link" generated by the
method.

From a larger perspective, having the "loaded_link" and
"no_results_link" properties in the I18N locale file is not ideal,
because it makes it difficult to configure these properties on
a per-server basis. For example, a "dev" machine might have searchers
configured to point to other development machines, while a "production"
machine would have searchers pointing to other production machines.
Since the I18N locale file can't be easily configured via server
environment variables, the file needs to be manually changed on either
the dev or production machine (depending on which version of the en.yml
file is checked into the Git repository).

The provided solution does the following:

1) Change the templates and code so that the "loaded_link" shown in the
   GUI comes exclusively from the "loaded_link" method of the searcher.

2) Change the templates and code so that a "no_results_link" method is
   available and used in the GUI.

3) For backward compatibility, preserve the existing behavior of
   retrieving "load_link" and "no_results_link" values from the
   I18N locale file, unless the properties in the file are nil, in
   which case use the appropriate method.

This would help in a number of ways:

1) It would make the code less confusing, and reduce duplication
   regarding the "loaded_link" instance.

2) It would enable the searcher to handle generating the links,
   including the ability to use more flexible mechanism such as
   environment variables, via the searcher configuration YAML files.

Note: Both "loaded_link" and "no_results_link" have associated "mobile"
versions, i.e. "loaded_link_mobile" and "no_results_link_mobile". The
"mobile" versions appear to be unused (or fall back to the "loaded_link"
property, if necessary), and so are not changed by the following.

----

Implementation Notes

In the existing implementation, the result of the "loaded_link" method
is displayed in the GUI when there is a successful search with results.
The "loaded_link" property from the I18N locale files is displayed when
a searcher error occurs.

The "loaded_link" property from the I18N locale file is used for
failed searches because the "search_all_in_threads" method of
SearcherConcern class returns an instance variable (named after the
searcher) that contains either the searcher object itself (if no error
occurs), or a StandardError (if an error occurs). The presence of
a StandardError in the instance variable triggers the appropriate
display in the view templates, but does not include a reference to the
searcher itself.

The code has been changed to enable the "loaded_link" method of the
searcher to be called even when there is an error. It does this by
wrapping the StandardError returned by the "search_all_in_threads"
method in a "SearcherError" class (which itself extends StandardError),
providing an accessor for retrieving the searcher. If an error has
occurred, the searcher is available to the ERB template, so that
the "loaded_link" method can be called.

In order to preserve the existing behavior of utilizing the
"loaded_link" property from the I18N locale file, logic has been placed
in the ERB templates to prefer the I18N property (if it exists) over
the "loaded_link" method of the searcher for failed searches.

For the "no_results_link", this implementation adds a "no_results_link"
method to the QuickSearch::Searcher base class. The class provides a
default implementation that preserves the existing behavior of
retrieving the "no_results_link" value from the I18N locale files, but
also adds the ability to retrieve the value from the searcher
configuration file.

kbeswick commented on app/views/quick_search/search/_module.html.erb:
```diff
@@ -1,15 +1,24 @@
 <div id="<%= service_name.dasherize %>" class="module-contents" >
     <% if searcher.is_a? StandardError %>
+
+        <% if I18n.exists?("#{service_name}_search.loaded_link") %>
```
My only suggestion on this pull request is to move this logic to a helper method, and set "module_link" by calling the helper method. It can also be used in _module_with_paging.html.erb.

kbeswick commented:
Thank you for the pull request, David. You raise good points re: loaded_link/no_result_link, I think it makes more sense to not be an i18n value so there can be more flexibility in how it is generated - I'm not sure how it came to be duplicated like that. 

I've tested with our implementation and a fresh demo QS site, everything seems to work as expected.

I've left one suggestion code-wise. If you could make that update, I'll merge this.

Thank you!

dsteelma-umd commented:
I've updated the code in line with your excellent suggestion. If I have misinterpreted what you were getting at, please let me know!

I've moved the "module_link" logic into a "module_link_on_error" class method in the QuickSearch::Searcher class. I used a class method so that the method could be called even if we don't know anything about the searcher (for example if we're still just getting a "StandardError" instead of a "SearcherError" from somewhere).

kbeswick commented:
Looks good. Thank you! Merged.
